pkg/api/v1/types.go

```diff
@@ -1217,12 +1217,14 @@ type EndpointSubset struct {
 	Ports     []EndpointPort    `json:"ports,omitempty" description:"port numbers available on the related IP addresses"`
 }
 
-// EndpointAddress is a tuple that describes single IP address.
+// EndpointAddress is a tuple that describes single endpoint address.
 type EndpointAddress struct {
-	// The IP of this endpoint.
-	// TODO: This should allow hostname or IP, see #4447.
+	// The IP of this endpoint, mutually exclusive with Hostname
 	IP string `json:"ip" description:"IP address of the endpoint"`
 
+	// Hostname for this endpoint, you can specify IP or Hostname, but not both.
+	Hostname string `json:"hostname" description:"Hostname for this endpoint"`
```

Comment:
> Would we set both IP and Hostname, or just OneOf?

In most places we really want one-of.  Yeah, making a previously required field suddenly be optional will potentially screw up clients who expect a field to be populated.  Changing it to be a hostname or IP is probably less likely to do damage, but it is sketchy to change.

LoadBalancerIngress was thrown together quickly.

HostPort and HostNet are named in equivalence with Docker.  One could argue it's a matter of perspective - Node is what we call a machine from outside that machine, Host is what we call it from within a container.  That's a stretch.

In this context though "hostname" (and HTTPGet.Host) means the name-or-IP of a host somewhere, and has nothing to do with kubernetes nodes.